lib/datasource.js

```diff
@@ -538,24 +525,112 @@ DataSource.prototype.defineRelations = function(modelClass, relations) {
       if ((targetModel && !isModelDataSourceAttached(targetModel)) ||
           (throughModel && !isModelDataSourceAttached(throughModel))) {
         // Create a listener to defer the relation set up
-        createListener(rn, r, targetModel, throughModel);
+        deferRelationSetup(relationName, r, targetModel, throughModel);
       } else {
         // The target model is resolved
         var params = traverse(r).clone();
-        params.as = rn;
-        params.model = polymorphicName || targetModel;
+        params.as = relationName;
+        params.model = targetModel;
         if (throughModel) {
           params.through = throughModel;
         }
-        modelClass[r.type].call(modelClass, rn, params);
+        modelClass[r.type].call(modelClass, relationName, params);
       }
     });
   }
 };
 
+function validateRelation(relationName, relation) {
+  var rn = relationName;
+  var r = relation;
+  var code;
+
+  // allows for method injection via relation definition in unit tests
+  getRelationValidationMsg = r.getRelationValidationMsg || getRelationValidationMsg;
+
```

Comment:
@ebarault I think we can catch the real error in tests, don't have to use the mock function, please check https://github.com/strongloop/loopback-datasource-juggler/pull/1301/files#r115806813, thanks.